alphacsc/_d_solver.py

```diff
@@ -183,122 +227,144 @@ def get_max_error_dict(self, z_encoder):
         [Yellin2017] BLOOD CELL DETECTION AND COUNTING IN HOLOGRAPHIC LENS-FREE
         IMAGING BY CONVOLUTIONAL SPARSE DICTIONARY LEARNING AND CODING.
         """
+        assert z_encoder.n_channels == self.n_channels
 
-        raise NotImplementedError()
+        d0 = z_encoder.get_max_error_patch()
 
-    def _init_windower(self, z_encoder):
-        raise NotImplementedError()
+        d0 = self._windower.window(d0)
 
-    def _get_objective(self, z_encoder):
+        return self.prox(d0)
 
-        def objective(D, full=False):
+    def init_dictionary(self, X):
+        """Returns a dictionary for the signal X depending on D_init value.
 
-            D = self.windower.window(D)
+        Parameter
+        ---------
+        X: array, shape (n_trials, n_channels, n_times)
+            The data on which to perform CSC.
 
-            return z_encoder.compute_objective(D)
+        Return
+        ------
+        D : array shape (n_atoms, n_channels + n_times_atom) or
+                  shape (n_atoms, n_channels, n_times_atom)
+            The initial atoms to learn from the data.
+        """
 
-        return objective
+        D_hat = self.dict_strategy.initialize(X)
 
+        if not hasattr(self.dict_strategy, 'D_init'):
+            D_hat = self._windower.window(D_hat)
 
-class Rank1DSolver(BaseDSolver):
-    """Base class for a rank1 solver d."""
+        self.D_hat = self.prox(D_hat)
 
-    def __init__(self, solver_d, uv_constraint, window, eps, max_iter,
-                 momentum, random_state, verbose, debug):
+        return self.D_hat
 
-        super().__init__(
-            solver_d, uv_constraint, window, eps, max_iter, momentum,
-            random_state, verbose, debug, rank1=True
-        )
+    def add_one_atom(self, z_encoder):
+        new_atom = self.get_max_error_dict(z_encoder)[0]
 
-        self.name = "Update uv"
+        self.D_hat = np.concatenate([self.D_hat, new_atom[None]])
 
-    def get_max_error_dict(self, z_encoder):
-        """Get the maximal reconstruction error patch from the data as a new atom
+        z_encoder.update_z_hat(self.D_hat)
```

Comment:
I would only use `set_D` and check that the number of component in `z_hat` are enough each time, without special case for `update_z_hat`